server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -147,149 +147,202 @@ public static IndexRequest getIndexWriteRequest(DocWriteRequest docWriteRequest)
 
     @Override
     protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<BulkResponse> listener) {
-        final long startTime = relativeTime();
-        final AtomicArray<BulkItemResponse> responses = new AtomicArray<>(bulkRequest.requests.size());
-
-        boolean hasIndexRequestsWithPipelines = false;
-        final MetaData metaData = clusterService.state().getMetaData();
-        ImmutableOpenMap<String, IndexMetaData> indicesMetaData = metaData.indices();
-        for (DocWriteRequest<?> actionRequest : bulkRequest.requests) {
-            IndexRequest indexRequest = getIndexWriteRequest(actionRequest);
-            if (indexRequest != null) {
-                // get pipeline from request
-                String pipeline = indexRequest.getPipeline();
-                if (pipeline == null) {
-                    // start to look for default pipeline via settings found in the index meta data
-                    IndexMetaData indexMetaData = indicesMetaData.get(actionRequest.index());
-                    // check the alias for the index request (this is how normal index requests are modeled)
-                    if (indexMetaData == null && indexRequest.index() != null) {
-                        AliasOrIndex indexOrAlias = metaData.getAliasAndIndexLookup().get(indexRequest.index());
-                        if (indexOrAlias != null && indexOrAlias.isAlias()) {
-                            AliasOrIndex.Alias alias = (AliasOrIndex.Alias) indexOrAlias;
-                            indexMetaData = alias.getWriteIndex();
-                        }
+        new BulkExecutor(task, bulkRequest, listener).run();
+    }
+
+    /**
+     * A runnable that will ensure the cluster state has been recovered enough to
+     * read index metadata and templates/pipelines.  Will retry up to the bulk's timeout
+     */
+    private final class BulkExecutor extends ActionRunnable<BulkResponse> {
```

Comment:
Do we need this class now that we aren't doing any retries ourself? I get that it wraps up the three parameters and the start time but 4 parameters to another method isn't so bad.